include/rmm/mr/host/new_delete_resource.hpp

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include "host_memory_resource.hpp"
+
+#include <cstddef>
+#include <utility>
+
+namespace rmm {
+namespace mr {
+
+/**---------------------------------------------------------------------------*
+ * @brief A `host_memory_resource` that uses the global `operator new` and
+ * `operator delete` to allocate host memory.
+ *---------------------------------------------------------------------------**/
+class new_delete_resource final : public host_memory_resource {
+ private:
+  /**---------------------------------------------------------------------------*
+   * @brief Allocates memory on the host of size at least `bytes` bytes.
+   *
+   * The returned storage is aligned to the specified `alignment` if supported,
+   * and to `alignof(std::max_align_t)` otherwise.
+   *
+   * @throws std::bad_alloc When the requested `bytes` and `alignment` cannot be
+   * allocated.
+   *
+   * @param bytes The size of the allocation
+   * @param alignment Alignment of the allocation
+   * @return void* Pointer to the newly allocated memory
+   *---------------------------------------------------------------------------**/
+  void* do_allocate(std::size_t bytes, std::size_t alignment =
+                                           alignof(std::max_align_t)) override {
+#if __cplusplus >= 201703L
+    return ::operator new(bytes, std::align_val_t(alignment));
+#else
+    // don't allocate anything if the user requested zero bytes
+    if (0 == bytes) {
+      return nullptr;
+    }
+    // allocate memory for bytes, plus potential alignment correction,
+    // plus store of the correction offset
+    void *p = ::operator new(bytes + alignment + sizeof(std::size_t));
+    std::size_t ptr_int = reinterpret_cast<std::size_t>(p);
+    // calculate the offset, i.e. how many bytes of correction were necessary
+    // to get an aligned pointer
+    std::size_t offset =
+        (ptr_int % alignment) ? (alignment - ptr_int % alignment) : 0;
+    // calculate the return pointer
+    char *ptr = static_cast<char *>(p) + offset;
+    // store the offset right after the actually returned value
+    std::size_t *offset_store = reinterpret_cast<std::size_t *>(ptr + bytes);
```

Comment:
Turns out this is actually undefined behavior. If `bytes % sizeof(size_t) != 0`, then this will result in a misaligned store (and misaligned load in the `do_deallocate`). 

Most CPU arches can handle misaligned load/stores (usually at a performance penalty), but according to the C++ standard it is UB. 

Instead, I'll store the offset immediately _prior_ to the returned pointer instead of at the end of the allocation.

Reply:
It's an arbitrary choice, but you could also align the location of the offset stored at the end of the allocation, right? Either one results in the same total storage?

Reply:
Assuming this is intended for host code only, I think it's a reasonable assumption that the platform that interacts with a CUDA-capable GPU supports unaligned load/stores. This is a known, finite set of platforms.

Reply:
> It's an arbitrary choice, but you could also align the location of the offset stored at the end of the allocation, right? Either one results in the same total storage?

I think it is arbitrary, yes, I was going to think through this tomorrow.

> Assuming this is intended for host code only, I think it's a reasonable assumption that the platform that interacts with a CUDA-capable GPU supports unaligned load/stores. This is a known, finite set of platforms.

Perhaps, but we should avoid UB when we can.